torch/_masked/__init__.py

```diff
@@ -390,21 +396,71 @@ def _canonical_dim(dim: DimOrDims, ndim: int) -> Tuple[int, ...]:
 
 def _input_mask(input: Tensor, *args, **kwargs) -> Tensor:
     """Return canonical input mask.
-    Canonical input mask is a boolean tensor with the same shape as
-    input and with (broadcasted) content of mask, if specified.
+
+    For strided inputs, the canonical input mask is a boolean tensor
+    with the same shape as input and with (broadcasted) content of
+    mask, if specified.
+
+    For sparse COO inputs, the canonical input mask is a sparse
+    coalesced tensor with the same shape and indices as the coalesced
+    :attr:`input` tensor.  The values of the canonical input masks are
+    all True, unless the :attr:`mask` tensor dtype is torch.bool, in
+    which case the :attr:`mask` values define the canonical input
+    mask.
+
+    If specified :attr:`mask` tensor uses a different layout from the
+    :attr:`input` tensor, then the canonical input mask tensor will
+    use the layout of the :attr:`input` tensor.
+
     """
+    if input.layout not in {torch.strided, torch.sparse_coo}:
+        raise ValueError(f'_input_mask expects strided or sparse COO tensor but got {input.layout}')
+
     mask = kwargs.get('mask')
+
+    # default mask
     if mask is None:
-        inmask = input.new_ones(input.shape, dtype=torch.bool)
-    elif mask.ndim < input.ndim:
-        inmask = torch.broadcast_to(mask.clone(), input.shape).to(dtype=torch.bool)
-    elif mask.ndim > input.ndim:
-        raise IndexError("_input_mask expected broadcastable mask (got mask dimensionality higher than of the input)")
-    elif mask.shape != input.shape:
-        inmask = torch.broadcast_to(mask.clone(), input.shape).to(dtype=torch.bool)
+        if input.layout == torch.strided:
+            mask = input.new_ones(input.shape, dtype=torch.bool)
+        else:
+            mask = input
+
+    # mask shape must match with input shape
+    if mask.shape != input.shape:
+        if mask.ndim > input.ndim:
+            raise IndexError("_input_mask expected broadcastable mask (got mask dimensionality higher than of the input)")
+        if mask.layout == torch.strided:
+            mask = torch.broadcast_to(mask.clone(), input.shape).to(dtype=torch.bool)
+        else:
+            mask = torch._sparse_broadcast_to(mask, input.shape)
+
+    # mask layout must match with input layout
+    if mask.layout != input.layout:
+        if input.layout == torch.strided:
+            if mask.dtype == torch.bool and not mask.is_coalesced():
+                # to_dense calls coalesce but coalesce not implemented for Bool
```

Comment:
Seems like a valuable side-PR again, since this will show up more and more

Reply:
I agree.

Reply:
Fixed via #74495